test/plugin/test_in_tail.rb

```diff
@@ -323,6 +350,69 @@ def test_emit_with_read_lines_limit(data)
+           "flat #{8192*10} bytes, 32 events"  => [:flat, 100, (8192 * 10), 32],
+           "flat #{8192*10} bytes, 32 events w/o stat watcher"  => [:flat_without_stat, 100, (8192 * 10), 32],
+           "parse #{8192*3} bytes, 48 events"  => [:parse, 100, (8192 * 3), 48],
+           "parse #{8192*3} bytes, 48 events w/o stat watcher"  => [:parse_without_stat, 100, (8192 * 3), 48],
+           "parse #{8192*10} bytes, 32 events" => [:parse, 100, (8192 * 10), 32],
+           "parse #{8192*10} bytes, 32 events w/o stat watcher" => [:parse_without_stat, 100, (8192 * 10), 32],
+           "flat 8k bytes with unit, 21 events"        => [:flat, 100, "8k", 21],
+           "flat 8k bytes with unit, 21 events w/o stat watcher"        => [:flat_without_stat, 100, "8k", 21],
+           "flat #{8*10}k bytes with unit, 32 events"  => [:flat, 100, "#{8*10}k", 32],
+           "flat #{8*10}k bytes with unit, 32 events w/o stat watcher"  => [:flat_without_stat, 100, "#{8*10}k", 32],
+           "parse #{8*3}k bytes with unit, 48 events"  => [:parse, 100, "#{8*3}k", 48],
+           "parse #{8*3}k bytes with unit, 48 events w/o stat watcher"  => [:parse_without_stat, 100, "#{8*3}k", 48],
+           "parse #{8*10}k bytes with unit, 32 events" => [:parse, 100, "#{8*10}k", 32],
+           "parse #{8*10}k bytes with unit, 32 events w/o stat watcher" => [:parse_without_stat, 100, "#{8*10}k", 32])
+      def test_emit_with_read_bytes_limit_per_second(data)
+        config_style, limit, limit_bytes, num_events = data
+        case config_style
+        when :flat
+          config = CONFIG_READ_FROM_HEAD + SINGLE_LINE_CONFIG + config_element("", "", { "read_lines_limit" => limit, "read_bytes_limit_per_second" => limit_bytes }) + CONFIG_MAX_THREAD_POOL_SIZE
+        when :parse
+          config = CONFIG_READ_FROM_HEAD + config_element("", "", { "read_lines_limit" => limit, "read_bytes_limit_per_second" => limit_bytes }) + PARSE_SINGLE_LINE_CONFIG + CONFIG_MAX_THREAD_POOL_SIZE
+        when :flat_without_stat
+          config = CONFIG_READ_FROM_HEAD + SINGLE_LINE_CONFIG + CONFIG_DISABLE_STAT_WATCHER + config_element("", "", { "read_lines_limit" => limit, "read_bytes_limit_per_second" => limit_bytes }) + CONFIG_MAX_THREAD_POOL_SIZE
+        when :parse_without_stat
+          config = CONFIG_READ_FROM_HEAD + CONFIG_DISABLE_STAT_WATCHER + config_element("", "", { "read_lines_limit" => limit, "read_bytes_limit_per_second" => limit_bytes }) + PARSE_SINGLE_LINE_CONFIG + CONFIG_MAX_THREAD_POOL_SIZE
+        end
+        d = create_driver(config)
+        msg = 'test' * 2000 # in_tail reads 8192 bytes at once.
+
+        d.run(expect_emits: 2) do
+          File.open("#{TMP_DIR}/tail.txt", "ab") {|f|
+            for _x in 0..20
+              f.puts msg
+            end
+          }
+        end
+
+        assert do
+          count_thread_pool_object >= 1
+        end
+
+        events = d.events
+        assert_true(events.length <= num_events)
```

Comment:
I think this test doesn't check this feature correctly, this assertion is too lough.
In addition, this feature seems cause hard time out while shutting down the plugin. The test shows following error:

```
E, [2021-04-05T18:47:53.212805 #298497] ERROR -- : unexpected error while after_shutdown, RuntimeError:event loop does NOT exit until hard timeout.
E, [2021-04-05T18:47:53.212914 #298497] ERROR -- : 	/home/aho/Projects/Fluentd/fluentd/lib/fluent/plugin_helper/event_loop.rb:130:in `after_shutdown'
E, [2021-04-05T18:47:53.212929 #298497] ERROR -- : 	/home/aho/Projects/Fluentd/fluentd/lib/fluent/test/driver/base.rb:156:in `block in instance_shutdown'
E, [2021-04-05T18:47:53.212940 #298497] ERROR -- : 	/home/aho/Projects/Fluentd/fluentd/lib/fluent/test/driver/base.rb:144:in `block in instance_shutdown'
E, [2021-04-05T18:47:53.212950 #298497] ERROR -- : 	/home/aho/Projects/Fluentd/fluentd/lib/fluent/test/driver/base.rb:156:in `instance_shutdown'
...
```

Reply:
> I think this test doesn't check this feature correctly, this assertion is too rough.

For example `num_events` of first data is `21` but it doesn't make sense because only 21 lines are inserted to the test file.

Reply:
> > I think this test doesn't check this feature correctly, this assertion is too rough.
> 
> For example `num_events` of first data is `21` but it doesn't make sense because only 21 lines are inserted to the test file.

I made testcases to be more robust:
https://github.com/fluent/fluentd/pull/3185/commits/f0228afa164343480336aba299dd3c956c10d0c9

Reply:
> In addition, this feature seems cause hard time out while shutting down the plugin. The test shows following error:
> 
> ```
> E, [2021-04-05T18:47:53.212805 #298497] ERROR -- : unexpected error while after_shutdown, RuntimeError:event loop does NOT exit until hard timeout.
> E, [2021-04-05T18:47:53.212914 #298497] ERROR -- : 	/home/aho/Projects/Fluentd/fluentd/lib/fluent/plugin_helper/event_loop.rb:130:in `after_shutdown'
> E, [2021-04-05T18:47:53.212929 #298497] ERROR -- : 	/home/aho/Projects/Fluentd/fluentd/lib/fluent/test/driver/base.rb:156:in `block in instance_shutdown'
> E, [2021-04-05T18:47:53.212940 #298497] ERROR -- : 	/home/aho/Projects/Fluentd/fluentd/lib/fluent/test/driver/base.rb:144:in `block in instance_shutdown'
> E, [2021-04-05T18:47:53.212950 #298497] ERROR -- : 	/home/aho/Projects/Fluentd/fluentd/lib/fluent/test/driver/base.rb:156:in `instance_shutdown'
> ...
> ```

I'm discussing with @cosmo0920 that this issue doesn't seems this PR's problem, it may be a same issue with #3323.